tools/make_patch.c

```diff
@@ -0,0 +1,944 @@
+		// Shitty hack, shouldn't be used anyway
+		int high = offset >> 8;
+		int low = offset & 0xFF;
+		fprintf(output, "0x%X%x", high, low);
+	} else if (strcmp(command, "Address") == 0 ||
+			   strcmp(command, "address") == 0) {
+		if (argc > 0) offset += strtol(argv[0], NULL, 0);
+
+		fprintf(output, "0x");
+		if (argc > 1) {
+			for (int i = strtol(argv[1], NULL, 0); i > 0; i--) {
+				putchar('0');
+			}
+		}
+
+		fprintf(output, isupper(command[0]) ? "%X" : "%x", offset);
+	} else if (strcmp(command, "Patch") == 0 ||
+			   strcmp(command, "patch") == 0) {
+		if (argc > 0) offset += strtol(argv[0], NULL, 0);
+
+		if (fseek(orig_rom, offset, SEEK_SET) != 0) {
+			fprintf(stderr, "Error: Could not seek to the offset of %s in the original ROM\n", currentpatch->name);
+			return;
+		}
+
+		if (fseek(new_rom, offset, SEEK_SET) != 0) {
+			fprintf(stderr, "Error: Could not seek to offset of %s in the new ROM\n", currentpatch->name);
+			return;
+		}
+
+		if (argc <= 1 || strcmp(argv[1], "big") != 0) {
+			int c = getc(new_rom);
+			if (c == getc(orig_rom)) {
+				fprintf(stderr, "Warning: %s doesn't actually contain any differences\n", currentpatch->name);
+			}
+
+			patch->offset = offset;
+			patch->size = 1;
+			fprintf(output, "0x");
+			fprintf(output, isupper(command[0]) ? "%02X" : "%02x", c);
+		} else {
+			int length = 0;
+
+			// Figure out the length of the patch
+			char *searchend = malloc(strlen(currentpatch->name) + 5);
+			strcpy(searchend, currentpatch->name);
+			strcat(searchend, "_End");
+			currentpatch = find_symbol(symbols, searchend);
+			length = currentpatch->value - offset;
+			memset(searchend, 0, (strlen(currentpatch->name)));
+
+			// We've got the length, now go back
+			fseek(new_rom, offset, SEEK_SET);
+
+			// Print out the patch
+			patch->offset = offset;
+			patch->size = length;
+			fprintf(output, "a%d:", length);
+			for (int i = 0; i < length; i++) {
+				if (i != 0) putc(' ', output);
+				fprintf(output, isupper(command[0]) ? "%02X" : "%02x", getc(new_rom));
+			}
+		}
+	} else if (strcmp(command, "Constant") == 0 ||
+			   strcmp(command, "constant") == 0) {
+		if (argc <= 1) {
+			fprintf(stderr, "Error: Missing arguments for %s", command);
+		}
+
+		int value = get_constant(argv[0], argv[1]);
+		if (value == -1) return;
+
+		fprintf(output, isupper(command[0]) ? "%02X %02X": "%02x %02x", value, value >> 8);
+	} else if (strcmp(command, "findaddress") == 0) {
+		if (argc != 1) {
+			fprintf(stderr, "Error: Missing argument for %s", command);
+		}
+		getsymbol = find_symbol(symbols, argv[0]);
+		if (!getsymbol) return;
+		fprintf(output, "0x%x", getsymbol->value);
+	} else if (strcmp(command, "conaddress") == 0) {
+		if (argc != 2) {
+			fprintf(stderr, "Error: Missing argument for %s", command);
+		}
+		getsymbol = find_symbol(symbols, argv[1]);
+		if (!getsymbol) return;
+		if (strcmp(argv[0], "dw") == 0) {
+			fprintf(output, "%02x ", (parse_offset(getsymbol->value, getsymbol->name[0]) % 0x100));
```

Comment:
`% 0x100` and `& 0xff` are basically the same; `/ 0x100` and `>> 8` are basically the same. We should have single macros that are consistently used in place of either:

```c
#define HIGH(x) (((x) >> 8) & 0xff)
#define LOW(x) ((x) & 0xff)
```

Reply:
Through preliminary testing.. I think HIGH should be:
```c
#define HIGH(x) ((x) >> 8)
```
Using your original suggestion messes up the patch file.

Reply:
Interesting; where is it doing `>> 8` on a value that's larger than 16-bit?

Reply:
I'll take a look at that tonight; I was testing your stuff this morning and didn't have time or was awake enough to really figure what why it wasn't working correctly.. I just changed it to what I posted above and it fixed the diff.

Reply:
Resolved in https://github.com/pret/pokecrystal/pull/882/commits/2bff40f5c9df301922cd8dda2745ead10ce85d6b and https://github.com/pret/pokecrystal/pull/882/commits/f63a3d5ddcf0bba45a2871d291b7a80a0f1550c5  

Also your original define HIGH(x) is working now... It was Probably something I messed up that morning.